drivers/hid/ipts/cmd.c

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Copyright (c) 2016 Intel Corporation
+ * Copyright (c) 2020-2022 Dorian Stoll
+ *
+ * Linux driver for Intel Precise Touch & Stylus
+ */
+
+#include <linux/errno.h>
+#include <linux/types.h>
+
+#include "context.h"
+#include "spec-device.h"
+
+static int ipts_cmd_get_errno(struct ipts_response rsp, enum ipts_status expect)
+{
+	if (rsp.status == IPTS_STATUS_SUCCESS)
+		return 0;
+
+	/*
+	 * If a status code was expected, dont produce an error.
+	 */
+	if (rsp.status == expect)
+		return 0;
+
+	/*
+	 * Some devices will always return this error. It is allowed
+	 * to ignore it and to try continuing.
+	 */
+	if (rsp.status == IPTS_STATUS_COMPAT_CHECK_FAIL)
+		return 0;
+
+	/*
+	 * Return something, this is not going to be checked.
+	 * Any error will just cause the driver to stop.
+	 */
+	return -EINVAL;
+}
+
+static int _ipts_cmd_recv(struct ipts_context *ipts, enum ipts_command_code code, void *data,
+			  size_t size, bool block, enum ipts_status expect)
+{
+	int ret;
+	struct ipts_response rsp = { 0 };
+
+	if (!ipts)
+		return -EFAULT;
+
+	if (size > sizeof(rsp.payload))
+		return -EINVAL;
+
+	if (block)
+		ret = mei_cldev_recv(ipts->cldev, (u8 *)&rsp, sizeof(rsp));
+	else
+		ret = mei_cldev_recv_nonblock(ipts->cldev, (u8 *)&rsp, sizeof(rsp));
+
+	if (ret == -EAGAIN)
+		return ret;
+
+	if (ret <= 0) {
+		dev_err(ipts->dev, "Error while reading response: %d\n", ret);
+		return ret;
+	}
+
+	/*
+	 * In a response, the command code will have the most significant bit
+	 * flipped to one. We check for this and then set the bit to 0.
+	 */
+	if ((rsp.cmd & 0x80000000) == 0) {
```

Comment:
Maybe consider adding a #define for that "magic value"?

Reply:
To be honest, I dont think we need a define for a simple bitmask, but changing it to `BIT(31)` would be a good idea.

Reply:
I mean you already have the `IPTS_CMD_...` stuff, why not add a `#define IPTS_CMD_RESPONSE_BIT BIT(31)` as well since you already use it twice? Reads a bit nicer IMHO and ties in a bit better to `spec-device.h` being a more complete representation of the spec.